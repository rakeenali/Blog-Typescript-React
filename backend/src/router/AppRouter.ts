import express from "express";

export default class AppRouter {
  static instance: express.Router;

  static getRouter(): express.Router {
    if (!AppRouter.instance) {
      AppRouter.instance = express.Router();
    }

    return AppRouter.instance;
  }
}
