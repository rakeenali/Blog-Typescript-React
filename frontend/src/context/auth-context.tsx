import * as React from "react";
import { navigate } from "@reach/router";

import { Dispatch, ContextProps } from "../types/types";

export interface IAuthState {
  isAuthenticated: boolean;
  token: string;
}

type Action = { type: "LOGOUT" } | { type: "LOGIN"; payload: string };

export const AuthStateContext = React.createContext<IAuthState | null>(null);
export const AuthDispatchContext = React.createContext<Dispatch<Action> | null>(
  null
);

const authReducer = (state: IAuthState, action: Action) => {
  switch (action.type) {
    case "LOGIN": {
      return { ...state, isAuthenticated: true, token: action.payload };
    }
    case "LOGOUT": {
      return { ...INITIALSTATE };
    }
    default: {
      throw new Error(`Action type is not found in user context`);
    }
  }
};

const INITIALSTATE: IAuthState = {
  isAuthenticated: false,
  token: ""
};

export const AuthProvider = ({ children }: ContextProps): JSX.Element => {
  const [state, dispatch] = React.useReducer(authReducer, INITIALSTATE);

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>
        {children}
      </AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  );
};

export function useAuthState(): IAuthState {
  const Context = React.useContext(AuthStateContext);

  if (!Context) {
    throw new Error(`useAuthState must be used iniside auth provider`);
  }

  return Context;
}

export function useAuthDispatch(): Dispatch<Action> {
  const Context = React.useContext(AuthDispatchContext);

  if (!Context) {
    throw new Error(`useAuthDispatch must be used iniside auth provider`);
  }

  return Context;
}

export const setUserAction = (
  token: string,
  dispatch: Dispatch<Action>
): void => {
  navigate("/profile");
  localStorage.setItem("blog:key", JSON.stringify(token));
  dispatch({ type: "LOGIN", payload: token });
};

export const getUserAction = (dispatch: Dispatch<Action>): void => {
  const rawToken = localStorage.getItem("blog:key");
  if (rawToken) {
    const token: string = JSON.parse(rawToken);
    dispatch({ type: "LOGIN", payload: token });
  }
};

export const logout = (dispatch: Dispatch<Action>): void => {
  localStorage.removeItem("blog:key");
  navigate("/login");
  dispatch({ type: "LOGOUT" });
};
