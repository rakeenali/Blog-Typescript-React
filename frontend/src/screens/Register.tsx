import React from "react";
import styled from "styled-components";
import axios from "axios";

import { Card, FormButton } from "../theme";
import Error from "../components/Error";
import { ErrorResponse, OKResponse } from "../types/types";
import { navigate } from "@reach/router";

interface IFormInput {
  username: string;
  password: string;
  name: string;
}

export default function Register(): JSX.Element {
  const [text, setText] = React.useState<IFormInput>({
    username: "",
    password: "",
    name: ""
  });

  const [showError, setShowError] = React.useState<boolean>(false);
  const [errorMessage, setErrorMessage] = React.useState<string>("");

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setText({ ...text, [e.target.name]: e.target.value });
  };

  const closeError = () => {
    setShowError(false);
    setErrorMessage("");
  };

  const registerUser = (e: React.FormEvent) => {
    e.preventDefault();
    axios
      .post("http://localhost:5000/auth/register", {
        name: text.name,
        password: text.password,
        username: text.username
      })
      .then((res: OKResponse) => {
        if (res && res.data.message) {
          navigate("/login");
        }
      })
      .catch((err: ErrorResponse) => {
        if (err && err.response) {
          setErrorMessage(err.response.data.error);
          setShowError(true);
        }
      });
  };

  return (
    <Center>
      {showError && <Error message={errorMessage} closeError={closeError} />}
      <Card height="480px" width="40%">
        <Inner>
          <Header>Register</Header>
          <div>
            <Form onSubmit={registerUser}>
              <FormControl>
                <label htmlFor="name">Name</label>
                <input
                  type="text"
                  id="name"
                  placeholder="Name"
                  onChange={onChange}
                  value={text.name}
                  required
                  name="name"
                />
              </FormControl>
              <FormControl>
                <label htmlFor="username">Username</label>
                <input
                  type="text"
                  id="username"
                  placeholder="Username"
                  onChange={onChange}
                  value={text.username}
                  required
                  name="username"
                />
              </FormControl>
              <FormControl>
                <label htmlFor="password">Password</label>
                <input
                  type="password"
                  id="password"
                  placeholder="Password"
                  onChange={onChange}
                  value={text.password}
                  required
                  name="password"
                />
              </FormControl>
              <FormControl>
                <FormButton>Submit</FormButton>
              </FormControl>
            </Form>
          </div>
        </Inner>
      </Card>
    </Center>
  );
}

const Center = styled.div`
  width: 100%;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Inner = styled.div`
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 10px;
`;

const Header = styled.h1`
  flex-basis: 20%;
  font-size: 50px;
`;
const Form = styled.form`
  flex-basis: 80%;
  height: 100%;
  width: 100%;
`;

const FormControl = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  margin-bottom: 20px;

  label {
    font-size: 20px;
    margin-bottom: 10px;
    text-align: center;
  }

  input {
    outline-color: red;
    padding: 10px;
    width: 25vw;
    background-color: #ccc;
    border-radius: 8px;
    border: none;
    font-size: 17px;
  }
`;
