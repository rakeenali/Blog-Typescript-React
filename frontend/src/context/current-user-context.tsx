import * as React from "react";
import { CurrentUserResponse as CurrentUserState } from "../types/inteface";
import { useAuthState } from "./auth-context";
import axios, { AxiosResponse } from "axios";
import { Dispatch, ContextProps } from "../types/types";

type Action = { type: "SET_USER"; payload: CurrentUserState };

export const CurrentUserStateContext = React.createContext<CurrentUserState | null>(
  null
);
export const CurrentUserDispatchContext = React.createContext<Dispatch<
  Action
> | null>(null);

const currentUserReducer = (state: CurrentUserState, action: Action) => {
  switch (action.type) {
    case "SET_USER": {
      return {
        ...state,
        ...action.payload
      };
    }
    default: {
      throw new Error(`Action type is not found in user context`);
    }
  }
};

const INITIALSTATE: CurrentUserState = {
  userId: null,
  name: null,
  username: null,
  usersProfile: null,
  profile: null,
  blogs: []
};

export const CurrentUserProvider = ({
  children
}: ContextProps): JSX.Element => {
  const [state, dispatch] = React.useReducer(currentUserReducer, INITIALSTATE);
  const authState = useAuthState();

  React.useLayoutEffect(() => {
    if (authState.isAuthenticated) {
      axios
        .get("http://localhost:5000/profile/current", {
          headers: { authorization: authState.token }
        })
        .then((res: AxiosResponse<CurrentUserState>) => {
          dispatch({ type: "SET_USER", payload: res.data });
        });
    }
  }, [authState]);

  return (
    <CurrentUserStateContext.Provider value={state}>
      <CurrentUserDispatchContext.Provider value={dispatch}>
        {children}
      </CurrentUserDispatchContext.Provider>
    </CurrentUserStateContext.Provider>
  );
};

export function useCurrentUserState(): CurrentUserState {
  const Context = React.useContext(CurrentUserStateContext);

  if (!Context) {
    throw new Error(
      `useCurrentUserState must be used iniside CurreuntUser provider`
    );
  }

  return Context;
}

export function useCurrentUserDispatch(): Dispatch<Action> {
  const Context = React.useContext(CurrentUserDispatchContext);

  if (!Context) {
    throw new Error(
      `useCurrentUserDispatch must be used iniside CurreuntUser provider`
    );
  }

  return Context;
}

export const getProfile = (token: string, dispatch: Dispatch<Action>): void => {
  axios
    .get("http://localhost:5000/profile/current", {
      headers: { authorization: token }
    })
    .then((res: AxiosResponse<CurrentUserState>) => {
      dispatch({ type: "SET_USER", payload: res.data });
    });
};
