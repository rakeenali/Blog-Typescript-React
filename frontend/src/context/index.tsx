import React from "react";

import { AuthProvider } from "./auth-context";

type AppProps = { children: React.ReactNode };

export default function AppProvider({ children }: AppProps): JSX.Element {
  return <AuthProvider>{children}</AuthProvider>;
}
