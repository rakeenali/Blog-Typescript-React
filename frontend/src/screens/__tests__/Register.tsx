import "@testing-library/jest-dom/extend-expect";

import * as React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import axios from "axios";
import { navigate } from "@reach/router";

import * as AuthContext from "../../context/auth-context";
import Register from "../Register";

jest.mock("@reach/router", () => {
  const navigate = jest.fn();
  return {
    navigate
  };
});

beforeEach(() => {
  jest.spyOn(axios, "post").mockImplementation((...args) => {
    console.warn("axios.post is not mocked for this call", ...args);
    return Promise.reject(new Error("This must be mocked!"));
  });
});

afterEach(() => {
  const mockNavigate = navigate as jest.Mock;
  const axiosPost = axios.post as jest.Mock;
  mockNavigate.mockClear();
  axiosPost.mockClear();
});

describe("<Register />", () => {
  test("renders correctly", () => {
    const { container } = render(
      <AuthContext.AuthProvider>
        <Register />
      </AuthContext.AuthProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test("render the form changes correctly", () => {
    const usernameText = "rakeen";
    const nameText = "Rakeen";
    const passwordText = "rakeen";

    const { getByPlaceholderText } = render(
      <AuthContext.AuthProvider>
        <Register />
      </AuthContext.AuthProvider>
    );

    const name = getByPlaceholderText("Name") as HTMLInputElement;
    const username = getByPlaceholderText("Username") as HTMLInputElement;
    const password = getByPlaceholderText("Password") as HTMLInputElement;

    fireEvent.change(name, {
      target: { value: nameText }
    });

    fireEvent.change(username, {
      target: { value: usernameText }
    });

    fireEvent.change(password, {
      target: { value: passwordText }
    });

    expect(username.value).toEqual(usernameText);
    expect(password.value).toEqual(passwordText);
    expect(name.value).toEqual(nameText);
  });

  test("register user successfuly", async () => {
    const usernameText = "rakeen";
    const nameText = "Rakeen";
    const passwordText = "rakeen";

    const axiosPost = axios.post as jest.Mock;

    const { getByPlaceholderText, queryByText } = render(
      <AuthContext.AuthProvider>
        <Register />
      </AuthContext.AuthProvider>
    );

    const name = getByPlaceholderText("Name") as HTMLInputElement;
    const username = getByPlaceholderText("Username") as HTMLInputElement;
    const password = getByPlaceholderText("Password") as HTMLInputElement;
    const submitButton = queryByText(/Submit/i);

    fireEvent.change(name, {
      target: { value: nameText }
    });

    fireEvent.change(username, {
      target: { value: usernameText }
    });

    fireEvent.change(password, {
      target: { value: passwordText }
    });

    axiosPost.mockResolvedValueOnce({
      status: 201,
      data: {
        message: "User created"
      }
    });

    submitButton && fireEvent.click(submitButton);

    await wait(
      () => {
        expect(navigate).toHaveBeenCalledTimes(1);
      },
      { timeout: 100 }
    );
  });

  test("invalid form submission throws error", async () => {
    const usernameText = "rakeen";
    const nameText = "Rakeen";

    const axisPost = axios.post as jest.Mock;

    const { getByPlaceholderText, queryByText } = render(
      <AuthContext.AuthProvider>
        <Register />
      </AuthContext.AuthProvider>
    );

    const name = getByPlaceholderText("Name") as HTMLInputElement;
    const username = getByPlaceholderText("Username") as HTMLInputElement;
    const password = getByPlaceholderText("Password") as HTMLInputElement;
    const submitButton = queryByText(/Submit/i);

    fireEvent.change(name, {
      target: { value: nameText }
    });

    fireEvent.change(username, {
      target: { value: usernameText }
    });

    fireEvent.change(password, {
      target: { value: "" }
    });

    axisPost.mockRejectedValueOnce({
      response: {
        data: {
          error: "Password field is required"
        }
      }
    });

    submitButton && fireEvent.click(submitButton);

    await wait(
      () => {
        expect(queryByText(/Password field is required/i)).toBeInTheDocument();
      },
      { timeout: 100 }
    );
  });
});
