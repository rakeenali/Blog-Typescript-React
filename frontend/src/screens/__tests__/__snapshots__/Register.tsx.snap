// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> renders correctly 1`] = `
<div
  class="sc-ifAKCX kMXFrU"
>
  <div
    class="sc-bdVaJa dpDiBc"
    height="480px"
    width="40%"
  >
    <div
      class="sc-EHOje ePbsYW"
    >
      <h1
        class="sc-bZQynM hQqxvj"
      >
        Register
      </h1>
      <div>
        <form
          class="sc-gzVnrw ieXoIT"
        >
          <div
            class="sc-htoDjs hJpmtM"
          >
            <label
              for="name"
            >
              Name
            </label>
            <input
              id="name"
              name="name"
              placeholder="Name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-htoDjs hJpmtM"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              id="username"
              name="username"
              placeholder="Username"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-htoDjs hJpmtM"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              id="password"
              name="password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
          </div>
          <div
            class="sc-htoDjs hJpmtM"
          >
            <button
              class="sc-bwzfXH ktHgFu"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
